### PR TITLE
API for batch withdrawal used in Swash

### PR DESCRIPTION
**Pull request by mabdi**
We have added three new APIs:

post `/auth/swash-register`:
sCollectors send swash users' wallet id (`address`) to this end point.
If it is a new user for adserver, it will register a new user for them and assign a new user id (`sAdId`).
If it is an existing user, the user id will be returned. 

post `/api/withdraw-swash-request`: 
An admin or moderator can invoke this API.
When this API is called, all users having a positive balance will be selected, and a withdrawal entry will be inserted on behalf of them.
All these entries have a common batch reference id.
The endpoint will return the batch reference id (`batch`).
It also runs a worker to transfer the sum of balances to an account owned by Swash in the BSC network

get `/api/withdraw-swash-status?batch=`:
After a reasonable wait, the moderator will call this endpoint, providing the batch id.
If the transaction is successful, the API will return the share of users (`shares`).
The moderator will receive the sum of balances in the BSC network and update Swash users' balances based on their shares.

### CODE REVIEW

File: app/Http/Controllers/Manager/AuthController.php [hunk 1]
```diff
@@ -418,6 +418,30 @@ public function walletLogin(Request $request): JsonResponse
...
+
+            $user = User::registerWithWallet($address, false);
+            $user->swash_wallet_address = $request->input('address');
+            if (Config::isTrueOnly(Config::AUTO_CONFIRMATION_ENABLED)) {
+                $this->confirmAdmin($user);
+            }
+            $user->saveOrFail();
+            DB::commit();
+        }
+        return response()->json(['sAd-id' => $user->wallet_address->toString()]);
```
Comment (aseyfpour) [inline app/Http/Controllers/Manager/AuthController.php:442]: I think it should be `sAdId`. In you test you check `sAdId`

File: app/Http/Controllers/Manager/WalletController.php [hunk 2]
```diff
@@ -328,6 +329,108 @@ public function withdraw(Request $request, AdsRpcClient $rpcClient): JsonRespons
         return $this->withdrawAds($request, $rpcClient);
     }
 
+    private function getSwashBSCAddress(): WalletAddress
```
Comment (aseyfpour) [inline app/Http/Controllers/Manager/WalletController.php:332]: I can't understand logic of this code, if it is a bad config you should raise error on running.
It is not a good time to validate swash bsc address on each request.

File: app/Http/Controllers/Manager/WalletController.php [hunk 3]
```diff
@@ -328,6 +329,108 @@ public function withdraw(Request $request, AdsRpcClient $rpcClient): JsonRespons
...
+                throw new InternalErrorException();
+            }
+        }
+        AdsSendOneSwash::dispatch(
+            $batchId,
+            $addressTo,
+            $amount,
+            $message
+        );
+        DB::commit();
```
Comment (aseyfpour) [inline app/Http/Controllers/Manager/WalletController.php:387]: What will happen if previous command works and send users money with executing the transaction but it fails to commit DB changes? Will you pay the users again in the next period?

File: app/Http/Controllers/Manager/WalletController.php [hunk 4]
```diff
@@ -328,6 +329,108 @@ public function withdraw(Request $request, AdsRpcClient $rpcClient): JsonRespons
...
+        /** @var User $user */
+        $user = Auth::user();
+        if(!$user->isModerator()){
+            return self::json(['Auth' => 'Only Moderator/Admin can run this API'], JsonResponse::HTTP_FORBIDDEN);
+        }
+        $batchId = $request->input('batch');
+        if (!$batchId){
+            return self::json([], Response::HTTP_BAD_REQUEST, ["param 'batch' is missing."]);
+        }
+        $userLedger = UserLedgerEntry::getFirstRecordByBatchId($batchId);
```
Comment (aseyfpour) [inline app/Http/Controllers/Manager/WalletController.php:403]: You havn't validate batch Id, what is the behavior of php if $userLedger be null in this case?

File: app/Jobs/AdsSendOneSwash.php [hunk 5]
```diff
@@ -0,0 +1,154 @@
...
+    /**
+     * Execute the job.
+     *
+     * @param AdsClient $adsClient
+     *
+     * @throws JobException
+     */
+    public function handle(AdsClient $adsClient): void
+    {
+        $userLedger = UserLedgerEntry::getFirstRecordByBatchId($this->batchId);
```
Comment (aseyfpour) [inline app/Jobs/AdsSendOneSwash.php:98]: Same as above

File: app/Models/User.php [hunk 6]
```diff
@@ -465,4 +470,11 @@ public static function fetchEmails(): Collection
     {
         return self::where('subscribe', 1)->whereNotNull('email')->get()->pluck('email');
     }
+
+    public static function generateRandomETHWalletForSwash(): string {
```
Comment (aseyfpour) [inline app/Models/User.php:474]: I think creating a normal wallet is more random than limiting chars and shortening the wallet length

File: app/Http/Controllers/Manager/WalletController.php (hunk 3 above)
Comment (mabdi) [reply app/Http/Controllers/Manager/WalletController.php:387]: The code is written based on `withdrawAds` line 504.
However, I swapped the commit and dispatch. I also added a log message to keep a track of the user shares.

File: app/Http/Controllers/Manager/AuthController.php (hunk 1 above)
Comment (mabdi) [reply app/Http/Controllers/Manager/AuthController.php:442]: Php automatically converts it to camel case style. 
However, I removed the dash and changed it to camel case style.

File: app/Http/Controllers/Manager/WalletController.php (hunk 2 above)
Comment (mabdi) [reply app/Http/Controllers/Manager/WalletController.php:332]: It's the first place this config is loaded.

File: app/Http/Controllers/Manager/WalletController.php (hunk 4 above)
Comment (mabdi) [reply app/Http/Controllers/Manager/WalletController.php:403]: Like. I return an error in the revised version.

File: app/Models/User.php (hunk 6 above)
Comment (mabdi) [reply app/Models/User.php:474]: Right. The leading zeros are removed.

File: app/Http/Controllers/Manager/WalletController.php [hunk 7]
```diff
@@ -328,6 +329,114 @@ public function withdraw(Request $request, AdsRpcClient $rpcClient): JsonRespons
...
+        }
+
+        $addressFrom = $this->getAdServerAdsAddress();
+        $address = $this->getSwashVaultBSCAddress();
+        $addressTo = $this->getWalletAdsAddress($rpcClient, $address);
+        $message = $this->getWalletAdsMessage($rpcClient, $address);
+        $batchId = hash('sha256', strval(microtime(true)) );
+
+        $balanceTotal = UserLedgerEntry::getWalletBalanceForAllUsers();
+        $users_balances = UserLedgerEntry::allWalletBalanceIfAny();
```
Comment (mabdi) [inline app/Http/Controllers/Manager/WalletController.php:358]: if `users_balances` is empty, return here. no need to go further.

File: app/Jobs/AdsSendOneSwash.php [hunk 8]
```diff
@@ -0,0 +1,157 @@
...
+        if (!$userLedger){
+            return;
```
Comment (mabdi) [inline app/Jobs/AdsSendOneSwash.php:100]: This situation means no record in the database corresponds to this batch id. So, return doesn't solve any problem. You should fail the job here.

File: app/Http/Controllers/Manager/WalletController.php [hunk 9]
```diff
@@ -328,6 +329,114 @@ public function withdraw(Request $request, AdsRpcClient $rpcClient): JsonRespons
...
+            return self::json(['Auth' => 'Only Moderator/Admin can run this API'], JsonResponse::HTTP_FORBIDDEN);
+        }
+
+        $addressFrom = $this->getAdServerAdsAddress();
+        $address = $this->getSwashVaultBSCAddress();
+        $addressTo = $this->getWalletAdsAddress($rpcClient, $address);
+        $message = $this->getWalletAdsMessage($rpcClient, $address);
+        $batchId = hash('sha256', strval(microtime(true)) );
+
+        $balanceTotal = UserLedgerEntry::getWalletBalanceForAllUsers();
```
Comment (mabdi) [inline app/Http/Controllers/Manager/WalletController.php:357]: If there is a non-swash user, you will send its balance to the side chain. Here, just get the balance for users having swash mapping.

File: app/Http/Controllers/Manager/WalletController.php [hunk 10]
```diff
@@ -328,6 +329,114 @@ public function withdraw(Request $request, AdsRpcClient $rpcClient): JsonRespons
...
+
+        if(!$user->isModerator()){
+            return self::json(['Auth' => 'Only Moderator/Admin can run this API'], JsonResponse::HTTP_FORBIDDEN);
+        }
+
+        $addressFrom = $this->getAdServerAdsAddress();
+        $address = $this->getSwashVaultBSCAddress();
+        $addressTo = $this->getWalletAdsAddress($rpcClient, $address);
+        $message = $this->getWalletAdsMessage($rpcClient, $address);
+        $batchId = hash('sha256', strval(microtime(true)) );
```
Comment (mabdi) [inline app/Http/Controllers/Manager/WalletController.php:355]: duplication possibility. If this line is executed twice in the same microsecond, the batch id will be equal!

File: app/Http/Controllers/Manager/WalletController.php [hunk 11]
```diff
@@ -328,6 +329,114 @@ public function withdraw(Request $request, AdsRpcClient $rpcClient): JsonRespons
...
+                -$item['share']
+            )->addressed($addressFrom, $addressTo);
+            if (!$ledgerEntry->save()) {
+                DB::rollBack();
+                throw new InternalErrorException();
+            }
+        }
+        DB::commit();
+        $logMessage = '[WalletController] Dispatching AdsSendOneSwash with batchId (%s) addressTo (%s) amount (%s) message (%s) users_balances (%s).';
+        Log::info(sprintf($logMessage, $batchId, $addressTo, $amount, $message, json_encode($users_balances)));
```
Comment (mabdi) [inline app/Http/Controllers/Manager/WalletController.php:383]: In the production, `info` logs normally are omitted. So, this logging makes no sense.

File: app/Jobs/AdsSendOneSwash.php [hunk 12]
```diff
@@ -0,0 +1,157 @@
...
+    /**
+     * Execute the job.
+     *
+     * @param AdsClient $adsClient
+     *
+     * @throws JobException
+     */
+    public function handle(AdsClient $adsClient): void
+    {
+        $userLedger = UserLedgerEntry::getFirstRecordByBatchId($this->batchId);
```
Comment (mabdi) [inline app/Jobs/AdsSendOneSwash.php:98]: If this line fails for any reason, the job will enter a loop. set a max attempt.

File: app/Http/Controllers/Manager/WalletController.php [hunk 13]
```diff
@@ -328,6 +329,114 @@ public function withdraw(Request $request, AdsRpcClient $rpcClient): JsonRespons
...
+        AdsSendOneSwash::dispatch(
+            $batchId,
+            $addressTo,
+            $amount,
+            $message
+        );
+        
+        $resp = array('code'=>  0, 'total'=> $amount, 'to' => config('app.svault_bsc_address'), 'batch'=>$batchId);
```
Comment (mabdi) [inline app/Http/Controllers/Manager/WalletController.php:391]: the JSON entries should be documented. use a `msg` as well.

File: app/Jobs/AdsSendOneSwash.php [hunk 14]
```diff
@@ -0,0 +1,157 @@
...
+            UserLedgerEntry::failAllRecordsInBatch($this->batchId, UserLedgerEntry::STATUS_SYS_ERROR);
+
+            Log::error(sprintf('[AdsSendOneSwash] Invalid txid: (%s)', $txid));
+
+            return;
+        }
+        UserLedgerEntry::acceptAllRecordsInBatch($this->batchId, $txid);
+    }
+
+    public function isTxValid(string $txid){
```
Comment (mabdi) [inline app/Jobs/AdsSendOneSwash.php:148]: why public?

File: config/app.php [hunk 15]
```diff
@@ -23,6 +25,13 @@
 $appEnv = env('APP_ENV', 'production');
 $aduserUrl = env('ADUSER_BASE_URL', env('ADUSER_INTERNAL_LOCATION', env('ADUSER_EXTERNAL_LOCATION')));
 
+try {
```
Comment (mabdi) [inline config/app.php:28]: Check it only in production.

File: database/migrations/2022_07_15_090824_add_batch_withdrawal_id_to_user_ledger_entries_table.php [hunk 16]
```diff
@@ -0,0 +1,32 @@
...
+{
+    /**
+     * Run the migrations.
+     *
+     * @return void
+     */
+    public function up()
+    {
+        Schema::table('user_ledger_entries', function (Blueprint $table) {
+            $table->char('batchid', 64)->nullable();
```
Comment (mabdi) [inline database/migrations/2022_07_15_090824_add_batch_withdrawal_id_to_user_ledger_entries_table.php:17]: add an index

File: routes/manager.php [hunk 17]
```diff
@@ -143,6 +143,8 @@ function () {
         Route::get('withdrawal-info', [WalletController::class, 'withdrawalInfo']);
         Route::post('calculate-withdrawal', [WalletController::class, 'calculateWithdrawal']);
         Route::post('wallet/withdraw', [WalletController::class, 'withdraw']);
+        Route::post('withdraw-swash', [WalletController::class, 'withdrawBatchSwash']);
```
Comment (mabdi) [inline routes/manager.php:146]: `withdraw-swash-reuqest`

File: routes/manager.php [hunk 18]
```diff
@@ -143,6 +143,8 @@ function () {
...
+        Route::get('withdraw-swash-info', [WalletController::class, 'withdrawSwashInfo']);
```
Comment (mabdi) [inline routes/manager.php:147]: `withdraw-swash-state`

**Pull request closed without merging**